### DISCUSSION TITLE
Add Representation.__rich_repr__() to support rich.pretty repr display, similar to __pretty__ from devtools

### DISCUSSION THREAD

**Discussion by StephenBrown2** [pydantic/pydantic#2606] [Feature Request]
There is a module that allows for incredible control over printing in the console, and also has a [pretty printing feature for repls](https://github.com/willmcgugan/rich#rich-repl), which can be enhanced by adding a `__rich_repl__()` function to iterate through the repl items, very similarly to the existing `__pretty__` function which is used by `devtools`. It doesn't handle indentation by default like `devtools`, but that can be configured.

Taking from the example in the docs https://pydantic-docs.helpmanual.io/usage/devtools/:
```py
from datetime import datetime
from typing import List

from pydantic import BaseModel


class Address(BaseModel):
    street: str
    country: str
    lat: float
    lng: float


class User(BaseModel):
    id: int
    name: str
    signup_ts: datetime
    friends: List[int]
    address: Address


user = User(
    id='123',
    name='John Doe',
    signup_ts='2019-06-01 12:22',
    friends=[1234, 4567, 7890],
    address=dict(street='Testing', country='uk', lat=51.5, lng=0),
)

from devtools import debug

print('\ndevtools.debug:\n')
debug(user)
print('\nshould be much easier read than:\n')
print('user:', user)

from rich import pretty, print
print('\nrich.pretty without __rich_repr__:\n')
pretty.pprint(user)

def rich_repr(self):
    for name, value in self.__repr_args__():
        yield name, value

from pydantic.utils import Representation

Representation.__rich_repr__ = rich_repr

print('\nwith __rich_repr__:\n')
pretty.pprint(user)
print('\nwith __rich_repr__ and expand_all=True:\n')
pretty.pprint(user, expand_all=True)
```
Screenshots for comparison:
![Screenshot 2021-03-30 at 12 59 18](https://user-images.githubusercontent.com/1148665/113042226-293dd600-9158-11eb-8103-f4a4505ce715.png)
![Screenshot 2021-03-30 at 12 59 41](https://user-images.githubusercontent.com/1148665/113042238-2c38c680-9158-11eb-92b9-a31bdbe19e2b.png)
![Screenshot 2021-03-30 at 13 00 01](https://user-images.githubusercontent.com/1148665/113042245-2d69f380-9158-11eb-8791-a9afd9bd3f71.png)

QUESTION: Would this be something you would be interested in adding, or should I submit an issue with rich to parse the existing `__pretty__` iterator output instead?

**Comment by willmcgugan** [comment]
Hi,

Author of Rich here.

I was going to suggest this myself at some point, but it might be premature now. I haven't documented `__rich_repr__` yet because I'm still trialing it. I wanted to run with it for a while to find the edge cases before looking to integrate it in to other libs.

Starting a discussion is great through. Here or in https://github.com/willmcgugan/rich/discussions

**Reply by StephenBrown2** [reply]
👋🏻  @willmcgugan awesome to see you here! I've really been enjoying using rich since I discovered it last year. I did see that the interface hasn't been documented yet, but I was able to figure out an MVP as noted above.

One thing that devtools' `__pretty__` does is automatic indentation, what do you think about incorporating that into `__rich_repr__` parsing?

**Reply by willmcgugan** [reply]
👋 

I think `__pretty__` has different (but overlapping) goals to `__rich_repr__`. In Rich I'm just interested in making a repr string that resembles how you might invoke it in code, but without any formatting. Rich has [another protocol](https://rich.readthedocs.io/en/latest/protocol.html) for formatting.

Your solution is good. It's nice how brief it is.

You can also yield three values if like -- the first two are name and value as before, the third is the default value. If the attribute is at the default then it will be omitted from the output.

@samuelcolvin do you have any thoughts on this?

**Reply by StephenBrown2** [reply]
> You can also yield three values if like -- the first two are name and value as before, the third is the default value. If the attribute is at the default then it will be omitted from the output.

I figured that would be possible, but wasn't sure how best to access the default value from within the meta-class. `ModelField` does return a `default` key in it's args: https://github.com/samuelcolvin/pydantic/blob/master/pydantic/fields.py#L950-L961, so it may be possible with a little logic to detect that and yield three if needed.

**Reply by StephenBrown2** [reply]
I also found that `rich` has an [inspect](https://rich.readthedocs.io/en/latest/introduction.html#rich-inspect) function which use may more closely align with `devtools.debug`, so I might just end up using that instead. Still would be nice to pretty-print Pydantic models after a
```py
from rich import pretty
pretty.install()
```
though.

**Reply by willmcgugan** [reply]
I've just added a PR for this. https://github.com/samuelcolvin/pydantic/pull/3099
